File: DoublyLinkedList.cpp
#include <iostream>
using namespace std;

class Node
{
public:
    int data;
    Node *next;
    Node *prev;
};

void insertAtStart(Node **head, int data)
{
    Node *temp = new Node();
    temp->data = data;
    temp->prev = NULL;
    temp->next = NULL;
    if (*head != NULL)
    {
        (*head)->prev = temp;
        temp->next = *head;
    }
    else
    {
        temp->next = NULL;
    }

    *head = temp;
}

void insertAtEnd(Node *head, int data)
{
    Node *temp = head;
    while (temp->next != NULL)
    {
        temp = temp->next;
    }
    Node *newNode = new Node();
    newNode->data = data;
    newNode->next = NULL;
    temp->next = newNode;
    newNode->prev = temp;
}

void insertAtMiddle(Node *head, int data, int n)
{
    Node *temp = head;
    for (int i = 1; i < n && temp; i++)
    {
        temp = temp->next;
    }
    Node *newNode = new Node();
    newNode->data = data;
    newNode->prev = temp;
    newNode->next = temp->next;
    temp->next = newNode;
}

void printList(Node *head)
{
    while (head != NULL)
    {
        cout << head->data << " ";
        head = head->next;
    }
    cout << endl;
}

int main()
{
    Node *head = NULL;
    insertAtStart(&head, 1);
    insertAtStart(&head, 2);
    insertAtStart(&head, 3);
    printList(head);

    insertAtEnd(head, 4);
    printList(head);

    insertAtMiddle(head, 5, 2);
    printList(head);
}

File: js/doublyLinkedList.js
class Node {
  constructor(data) {
    this.data = data;
    this.next = null;
    this.prev = null;
  }
}

class LinkedList {
  constructor() {
    this.head = null;
    this.size = 0;
  }

  add(element) {
    var newNode = new Node(element);
    if (this.head == null) {
      this.head = newNode;
    } else {
      var current = this.head;
      while (current) {
        current = current.next;
      }
      current.next = newNode;
      newNode.prev = current;
      newNode.next = null;
    }
    this.size++;
  }

  delete(element) {
    var current,
      previous,
      i = 0;
    if (this.head.data == element) {
      this.head = this.head.next;
      this.head.prev = null;
    } else {
      while (current) {
        if (current.data == element) {
          previous.next = current.next;
          current.prev = previous;
          this.size--;
          return i;
        }
        i++;
        previous = current;
        current = current.next;
      }
    }
    return -1;
  }

  printList() {
    var current = this.head;
    var str = "";
    while (current) {
      str += current.data + " ";
      current = current.next;
    }
    console.log(str);
  }
}

var ll = new LinkedList();
ll.add(10);
ll.add(20);
ll.add(30);
ll.add(40);
ll.add(50);
console.log("Removed: ", ll.remove(30));
ll.printList();
